From 4b15f5e8766dcd709728c665280cad84d29e4222 Mon Sep 17 00:00:00 2001
From: zhangyichix <yichix.zhang@intel.com>
Date: Mon, 10 Oct 2022 06:04:24 +0000
Subject: [PATCH] Exit the loop when the surface may not be available

cases:
android.media.decoder.cts.DecoderTest#testCodecResetsH264WithSurface
android.media.decoder.cts.DecoderTest#testCodecResetsHEVCWithSurface

From Android T, Surfaces are no longer used when MediaCodec#stop() is called.
https://android-review.googlesource.com/c/platform/frameworks/av/+/2098075
Exit the loop when the surface may not be available.

Tracked-On: OAM-104001
Signed-off-by: zhangyichix <yichix.zhang@intel.com>
---
 c2_components/include/mfx_c2_decoder_component.h |  9 +++++++++
 c2_components/src/mfx_c2_decoder_component.cpp   | 13 ++++++++++++-
 2 files changed, 21 insertions(+), 1 deletion(-)

diff --git a/c2_components/include/mfx_c2_decoder_component.h b/c2_components/include/mfx_c2_decoder_component.h
index d5842cc..1e925e5 100755
--- a/c2_components/include/mfx_c2_decoder_component.h
+++ b/c2_components/include/mfx_c2_decoder_component.h
@@ -43,6 +43,13 @@ public:
         DECODER_AV1,
     };
 
+    enum class OperationState {
+        INITIALIZATION,
+        RUNNING,
+        STOPPING,
+        STOPPED,
+    };
+
 protected:
     MfxC2DecoderComponent(const C2String name, const CreateConfig& config,
         std::shared_ptr<C2ReflectorHelper> reflector, DecoderType decoder_type);
@@ -174,6 +181,8 @@ private:
 
     bool m_bInitialized;
 
+    OperationState m_OperationState { OperationState::INITIALIZATION };
+
     MfxCmdQueue m_workingQueue;
     MFX_TRACEABLE(m_workingQueue);
     MfxCmdQueue m_waitingQueue;
diff --git a/c2_components/src/mfx_c2_decoder_component.cpp b/c2_components/src/mfx_c2_decoder_component.cpp
index 55b071c..d212997 100755
--- a/c2_components/src/mfx_c2_decoder_component.cpp
+++ b/c2_components/src/mfx_c2_decoder_component.cpp
@@ -738,13 +738,15 @@ c2_status_t MfxC2DecoderComponent::DoStart()
 
     } while(false);
 
+    m_OperationState = OperationState::RUNNING;
+
     return C2_OK;
 }
 
 c2_status_t MfxC2DecoderComponent::DoStop(bool abort)
 {
     MFX_DEBUG_TRACE_FUNC;
-
+    m_OperationState = OperationState::STOPPING;
     // Working queue should stop first otherwise race condition
     // is possible when waiting queue is stopped (first), but working
     // queue is pushing tasks into it (via DecodeFrameAsync). As a
@@ -780,6 +782,7 @@ c2_status_t MfxC2DecoderComponent::DoStop(bool abort)
         m_c2Bitstream->GetFrameConstructor()->Close();
     }
 
+    m_OperationState = OperationState::STOPPED;
     return C2_OK;
 }
 
@@ -1740,6 +1743,14 @@ c2_status_t MfxC2DecoderComponent::AllocateC2Block(uint32_t width, uint32_t heig
             break;
         }
 
+        // From Android T, Surfaces are no longer used when MediaCodec#stop() is called.
+        // https://android-review.googlesource.com/c/platform/frameworks/av/+/2098075
+        // Exit the loop when the surface may not be available.
+        if (OperationState::STOPPING == m_OperationState) {
+            res = C2_CANCELED;
+            break;
+        }
+
         if (m_mfxVideoParams.IOPattern == MFX_IOPATTERN_OUT_VIDEO_MEMORY) {
 
             C2MemoryUsage mem_usage = {m_consumerUsage, C2AndroidMemoryUsage::HW_CODEC_WRITE};
-- 
2.40.0

